src/mesh.cpp

```diff
@@ -1373,6 +1403,442 @@ bool RectilinearMesh::intersects(Position& r0, Position r1, int* ijk) const
+      p_max_split = std::stoi(get_node_value(photon_wwp,"max_split"));
+      if (p_max_split<=1 )   fatal_error("Max split number must bigger than 1.");
+    }
+      
+    // multiplier for weight window lower bounds
+    if (check_for_node(photon_wwp, "multiplier")) {
+      p_multiplier = std::stod(get_node_value(photon_wwp,"multiplier"));
+      if (p_multiplier<=0 )   fatal_error("Multiplier for lower weight window must bigger than 0.");
+    }     
+  }
+  // photon
+    
+  // user defined source weight biasing in energy
+  if (check_for_node(node, "user_defined_biasing")) {
+    user_defined_biasing = true;
+    xml_node node_user_defined_biasing = node.child("user_defined_biasing");
+    
+    // energy group for source weight biasing
+    if (check_for_node(node_user_defined_biasing, "biasing_energy")) {
+      auto value = get_node_xarray<double>(node_user_defined_biasing, "biasing_energy");
+      biasing_energy.push_back(0.);
+      for (int i=0; i<value.size(); i++)   biasing_energy.push_back(value.at(i));
+    } else {
+      fatal_error("Must provide energy group for biasing.");
+    }
+
+    // origin possibility for each energy group
+    if (check_for_node(node_user_defined_biasing, "origin_possibility")) {
+      auto value = get_node_xarray<double>(node_user_defined_biasing, "origin_possibility");
+      if (value.size()!=biasing_energy.size()-1)  fatal_error("Origin_possibility and biasing_energy must have the same number of input.");
+      origin_possibility.push_back(0.);
+      cumulative_possibility.push_back(0.);
+      for (int i=0; i<value.size(); i++) {
+        origin_possibility.push_back(value.at(i));
+        cumulative_possibility.push_back(0.);
+      }
+      // normalization
+      double total_possibility=0.0;
+      for (int i=0; i<origin_possibility.size(); i++)   total_possibility += origin_possibility.at(i);
+      for (int i=1; i<origin_possibility.size(); i++)  {   
+        origin_possibility.at(i) = origin_possibility.at(i) / total_possibility;
+        cumulative_possibility.at(i) = cumulative_possibility.at(i-1) + origin_possibility.at(i);
+      }
+    } else {
+      fatal_error("Must provide origin_possibility for each group.");
+    }
+      
+    // biasing weight for each energy group
+    if (check_for_node(node_user_defined_biasing, "biasing")) {
+      auto value = get_node_xarray<double>(node_user_defined_biasing, "biasing");
+      if (value.size()!=biasing_energy.size()-1)  fatal_error("Biasing and biasing_energy must have the same number of input.");
+      biasing.push_back(0);
+      cumulative_biasing.push_back(0);
+      for (int i=0; i<value.size(); i++) {
+        biasing.push_back(value.at(i));
+        cumulative_biasing.push_back(0);
+      }
+      // normalization
+      double total_possibility=0.0;
+      for (int i=0; i<biasing.size(); i++)   total_possibility += biasing.at(i);
+      for (int i=1; i<biasing.size(); i++) {
```

Comment:
```suggestion
      for (int i = 0; i < biasing.size(); i++) total_possibility += biasing.at(i);
      for (int i = 1; i < biasing.size(); i++) {
```
I'd consider using one of the standard library algorithms to get the total probability, like `std::accumulate`.

Reply:
I don't know the `std::accumulate` before. I may try to change this part later.

Reply:
I'm not sure if `std::accumulate` will be faster. I need to calculate the cumulative probability for each energy bin here. So if I use `std::accumulate` ，I need to call this function for each energy bin, right? In this way, I thought it may be slower than calculating the cumulative probability directly. I'm not sure if I understood you well, so please correct me if I was wrong.

Reply:
Hrrmm it honestly may not be much faster. In my mind it might be nice from a readability standpoint. I was thinking the lines 1818-1819 could be combined into:

```cpp
double total_possibility = std::accumulate(biasing.begin(), biasing.end(), 0.0);
```
Maybe I'm missing something though.

Reply:
I didn't check my code carefully yesterday and misunderstood my code with my memory.  I changed the code like this, I think this should be supposed to what you are looking forward.
```
      double total_probability = std::accumulate(biasing.begin(), biasing.end(), 0.0);
      for (int i = 1; i < biasing.size(); i++) {
        biasing.at(i) = biasing.at(i)/total_probability;
        cumulative_biasing.at(i) = cumulative_biasing.at(i-1) + biasing.at(i);
      }
```